### PR TITLE
ci: avoid using same port number for different net tests

### PR DESCRIPTION
**Pull request by keroxp**
Some net tests (cli,std) are unstable in any platform. I doubt that one of causes is they are using same port numbers during tests.  I added port picker that iterates different port number in private port range (49152 - 65535) for cli and std.

And note that when `cargo test` is running, we are using addresses  below by `tools/http_server.ts`:

```
PORT = 4545
REDIRECT_PORT = 4546
ANOTHER_REDIRECT_PORT = 4547
DOUBLE_REDIRECTS_PORT = 4548
INF_REDIRECTS_PORT = 4549
HTTPS_PORT = 5545
```

I also doubt this is the main cause of windows's too flaky CI on net tests. 🤔

### CODE REVIEW

**Comment by keroxp** [comment]
#4107 is related to this.

**Comment by keroxp** [comment]
@ry How do you think?

File: cli/js/tests/test_util.ts [hunk 1]
```diff
@@ -398,3 +398,21 @@ unitTest(
...
+    yield i;
+    i++;
+    if (i > 55000) {
+      i = 55000;
+    }
+  }
+}
+const it = portIterator();
+/** Obtain (maybe) safe port number for net tests */
+export function usePort(): number {
```
Comment (bartlomieju) [inline cli/js/tests/test_util.ts]: Please rename to `randomPort`

File: std/http/internal/test_util.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { assert } from "../../testing/asserts.ts";
```
Comment (bartlomieju) [inline std/http/internal/test_util.ts]: No need to put it inside `internal/` directory, please lift one level up

File: std/http/internal/test_util.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    yield i;
+    i++;
+    if (i > 65535) {
+      i = 55001;
+    }
+  }
+}
+const it = portIterator();
+/** Obtain (maybe) safe port number for net tests */
+export function usePort(): number {
```
Comment (bartlomieju) [inline std/http/internal/test_util.ts]: s/usePort/randomPort/

**Comment by nayeemrmn** [comment]
Are you sure this won't hide bugs? If resources are cleaned properly it should be okay to use the same port. Or maybe it can't be helped due to shutdowns not working properly on Windows, or something?

**Comment by keroxp** [comment]
@nayeemrmn No, thank you for pointing. I know tests around network are very unstable on windows. But what you concern about should be tested by one for that.

The motivation to use random port is that there are no rules to pick port from host and sometimes it conflicts existing local services. Port number between 1000 ~ 10000 are often taken by developer. It's slightly annoying to stop other services before running Deno's test. 

Of course,  that doesn't mean those ports are safe.
Basically 49152 (+20~30) for js/cli, 55001 (+20 ~ 30) for std will be used.

**Comment by keroxp** [comment]
@bartlomieju Thank you for reviewing. I've applied reviews, please check!

**Comment by ry** [comment]
This should wait until #4373 is fixed.

I'm hesitant on this patch because it complicates the test setup without a clear benefit.

Sorry for the delayed review @keroxp ...

File: std/http/test_util.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    }
+  }
+}
+const it = portIterator();
+/** Obtain (maybe) safe port number for net tests */
+export function randomPort(): number {
+  const { value } = it.next();
+  assert(value != null);
+  return value;
+}
```
Comment (ry) [inline std/http/test_util.ts]: This could be done more simply without an iterator and a global variable...
